YakimS/assignment2/BinaryTree.java

```diff
@@ -124,108 +71,83 @@ else if(printAncestors(key,root.getLeftSon()) ||
-     * each given key, than return the last common number in both paths.
+     * This function find the lowest common ancestor by traversing up the tree
+     * of each given node till reaching a common ancestor, which is the lowest
      *
-     * @param key1 - first key to find common ancestor
-     * @param key2 - second key to find common ancestor
-     * @return lowest common ancestor of the two given keys. If the keys are equal than
-     *          the function will return the key number.
-     * @throws KeyNotFoundException if one or both of the keys are not in the tree
+     * @param value1 - first value to find common ancestor
+     * @param value2 - second value to find common ancestor
+     * @return lowest common ancestor of the two given values. If the values are equal than
+     *          the function will return the value number.
+     * @throws IllegalArgumentException if one or both of the values are not in the tree
      */
-    public int getLowestCommonAncestor(int key1, int key2) throws KeyNotFoundException {
+    public T getLowestCommonAncestor(T value1, T value2) throws IllegalArgumentException {
 
-        // create paths to the two keys given
-        List<Integer> firstKeyPath = getKeyPath(key1, root, new LinkedList<>());
-        List<Integer> secKeyPath = getKeyPath(key2, root, new LinkedList<>());
+        TreeNode<T> nodeInPathOfVal1 = findNodeByVal(value1, root);
+        TreeNode<T> nodeInPathOfVal2 = findNodeByVal(value2, root);
 
-        if(firstKeyPath == null || secKeyPath == null)
+        if(nodeInPathOfVal1 == null || nodeInPathOfVal2 == null)
         {
-            throw new KeyNotFoundException();
+            throw new IllegalArgumentException("Value not found");
+        }
+        else if(nodeInPathOfVal1.equals(nodeInPathOfVal2))
+        {
+            return nodeInPathOfVal1.getValue();
+        }
+        else if(nodeInPathOfVal1.equals(root) || nodeInPathOfVal2.equals(root))
+        {
+            return root.getValue();
         }
 
-        Collections.reverse(firstKeyPath);
-        Collections.reverse(secKeyPath);
-
-        // iterate throw the paths till they are stop intersecting - this will
-        // be where the lowest common ancestor at
-        int currentLowestCommonAncestor = root.key;
-        Iterator IteratorFirst = firstKeyPath.iterator();
-        Iterator IteratorSecond = secKeyPath.iterator();
-        while (IteratorFirst.hasNext() && IteratorSecond.hasNext()) {
-            int nextInFirstPath = (int) IteratorFirst.next();
-            int nextInSecondPath = (int) IteratorSecond.next();
-            if(nextInFirstPath == nextInSecondPath)
+        while (nodeInPathOfVal1.getLevel() != 0 && nodeInPathOfVal2.getLevel()!=0)
+        {
+            if(nodeInPathOfVal1.equals(nodeInPathOfVal2))
+            {
+                return nodeInPathOfVal1.getValue();
+            }
+            else if(nodeInPathOfVal1.getLevel() < nodeInPathOfVal2.getLevel())

```

Comment:
Optional: No need for `else` here. Having two separate ifs (one for node equality, one for level calculations) will be a bit easier to read, since they are logically separate.